File: translation-library/python/src/translation_library/translators/yandex.py
"""
Переводчик Yandex Translate
Основан на реализации из SillyTavern
"""

import uuid
from typing import Optional

import requests

from .base import BaseTranslator
from ..types import (
    TranslationOptions,
    TranslationResult,
    TranslationError,
    TranslatorProvider,
)
from ..chunker import chunk_with_links, restore_text_with_links


class YandexTranslator(BaseTranslator):
    """Переводчик Yandex Translate"""
    
    def __init__(
        self,
        api_key: Optional[str] = None,
        endpoint: Optional[str] = None,
        timeout: int = 30000,
        retries: int = 3,
        chunk_size: int = 5000,
    ):
        super().__init__(
            provider=TranslatorProvider.YANDEX,
            api_key=api_key,
            endpoint=endpoint,
            timeout=timeout,
            retries=retries,
        )
        self.chunk_size = chunk_size
    
    def translate(self, text: str, options: TranslationOptions) -> TranslationResult:
        """
        Выполняет перевод текста с использованием Yandex Translate
        
        Args:
            text: Текст для перевода
            options: Опции перевода
            
        Returns:
            Результат перевода
        """
        self.validate_input(text, options.target_language)
        
        target_lang = self.normalize_language_code(options.target_language)
        
        try:
            # Разделяем текст на чанки для обработки
            chunks, links = chunk_with_links(text, self.chunk_size)
            
            translated_chunks = []
            
            for chunk in chunks:
                if not chunk.strip():
                    continue
                
                translated_chunk = self._translate_chunk(chunk, target_lang)
                translated_chunks.append(translated_chunk)
            
            # Восстанавливаем текст с сохранением ссылок
            translated_text = restore_text_with_links(translated_chunks, links)
            
            return TranslationResult(
                text=translated_text,
                source_language=None,  # Yandex определяет язык автоматически
                provider=self.provider,
            )
        except TranslationError:
            raise
        except Exception as error:
            raise self.create_error(
                f'Yandex Translate translation failed: {str(error)}',
                details=error
            )
    
    def _translate_chunk(self, chunk: str, target_lang: str) -> str:
        """
        Переводит один чанк текста
        
        Args:
            chunk: Текст для перевода
            target_lang: Целевой язык
            
        Returns:
            Переведённый текст
        """
        # Используем публичный API Yandex без API ключа, как в SillyTavern
        url = 'https://translate.yandex.net/api/v1/tr.json/translate'
        
        # Генерируем уникальный идентификатор
        ucid = str(uuid.uuid4()).replace('-', '')
        
        params = {
            'ucid': ucid,
            'srv': 'android',
            'format': 'text',
            'text': chunk,
            'lang': target_lang,
        }
        
        try:
            response = requests.post(url, data=params, timeout=self.timeout / 1000)
            
            if response.status_code != 200:
                raise self.create_error(
                    f'Yandex Translate API error: {response.status_code} - {response.text}',
                    status_code=response.status_code,
                )
            
            result = response.json()
            
            if 'code' in result and result['code'] != 200:
                raise self.create_error(
                    f'Yandex Translate API error: {result.get("message", "Unknown error")}',
                    status_code=result.get('code'),
                    details=result
                )
            
            if 'text' not in result:
                raise self.create_error(
                    'Invalid response from Yandex Translate API',
                    details=result
                )
            
            return result['text'][0]
        except requests.RequestException as error:
            raise self.create_error(
                f'Network error: {str(error)}',
                details=error
            )
    
    def detect_language(self, text: str) -> str:
        """
        Определяет язык текста
        
        Args:
            text: Текст для определения языка
            
        Returns:
            Код языка
        """
        # Используем публичный API Yandex без API ключа, как в SillyTavern
        url = 'https://translate.yandex.net/api/v1/tr.json/translate'
        
        # Генерируем уникальный идентификатор
        ucid = str(uuid.uuid4()).replace('-', '')
        
        params = {
            'ucid': ucid,
            'srv': 'android',
            'format': 'text',
            'text': text,
            'lang': 'auto',
        }
        
        try:
            response = requests.post(url, data=params, timeout=self.timeout / 1000)
            
            if response.status_code != 200:
                raise self.create_error(
                    f'Yandex Translate API error: {response.status_code} - {response.text}',
                    status_code=response.status_code,
                )
            
            result = response.json()
            
            if 'code' in result and result['code'] != 200:
                raise self.create_error(
                    f'Yandex Translate API error: {result.get("message", "Unknown error")}',
                    status_code=result.get('code'),
                    details=result
                )
            
            if 'lang' not in result:
                raise self.create_error(
                    'Invalid response from Yandex Translate API',
                    details=result
                )
            
            return result['lang'].split('-')[0]
        except requests.RequestException as error:
            raise self.create_error(
                f'Network error: {str(error)}',
                details=error
            )
    
    def get_supported_languages(self) -> list[str]:
        """
        Получает список поддерживаемых языков
        
        Returns:
            Список языков
        """
        return [
            'af', 'sq', 'am', 'ar', 'hy', 'az', 'be', 'bn', 'bg', 'ca',
            'ceb', 'zh-CN', 'co', 'cs', 'da', 'nl', 'en', 'eo', 'et',
            'fi', 'fr', 'fy', 'gl', 'ka', 'de', 'el', 'gu', 'ht', 'ha',
            'haw', 'he', 'hi', 'hmn', 'hu', 'is', 'ig', 'id', 'ga', 'it',
            'ja', 'jv', 'kn', 'kk', 'km', 'rw', 'ko', 'ku', 'ky', 'lo',
            'la', 'lv', 'lt', 'lb', 'mk', 'mg', 'ms', 'ml', 'mt', 'mi',
            'mr', 'mn', 'my', 'ne', 'no', 'ny', 'or', 'ps', 'fa', 'pl',
            'pt', 'pa', 'ro', 'ru', 'sm', 'gd', 'sr', 'st', 'sn', 'sd',
            'si', 'sk', 'sl', 'so', 'es', 'su', 'sw', 'sv', 'tg', 'ta',
            'tt', 'te', 'th', 'tr', 'tk', 'uk', 'ur', 'ug', 'uz', 'vi',
            'cy', 'xh', 'yi', 'yo', 'zu',
        ]

File: translation-library/nodejs/src/translation-library/translators/yandex.ts
/**
 * Переводчик Yandex Translate
 * Основан на реализации из SillyTavern
 */

import fetch from 'node-fetch';
import { v4 as uuidv4 } from 'uuid';
import { BaseTranslator } from './base';
import {
  TranslationOptions,
  TranslationResult,
  TranslationError,
  TranslatorProvider,
} from '../types';
import { chunkText, restoreTextWithLinks, chunkWithLinks } from '../chunker';

/**
 * Переводчик Yandex Translate
 */
export class YandexTranslator extends BaseTranslator {
  private readonly apiUrl = 'https://translate.yandex.net/api/v1/tr.json/translate';
  private readonly chunkSize: number = 5000;

  constructor(config: {
    apiKey?: string;
    endpoint?: string;
    timeout?: number;
    retries?: number;
    chunkSize?: number;
  } = {}) {
    super({
      provider: 'yandex',
      apiKey: config.apiKey,
      endpoint: config.endpoint,
      timeout: config.timeout,
      retries: config.retries,
    });

    this.chunkSize = config.chunkSize || this.chunkSize;
  }

  /**
   * Выполняет перевод текста с использованием Yandex Translate
   * @param text Текст для перевода
   * @param options Опции перевода
   * @returns Результат перевода
   */
  async translate(text: string, options: TranslationOptions): Promise<TranslationResult> {
    this.validateInput(text, options.targetLanguage);

    const targetLang = this.normalizeLanguageCode(options.targetLanguage);

    try {
      // Разделяем текст на чанки для обработки
      const { chunks, links } = chunkWithLinks(text, this.chunkSize);

      const translatedChunks: string[] = [];

      for (const chunk of chunks) {
        if (chunk.trim().length === 0) {
          continue;
        }

        const translatedChunk = await this.translateChunk(chunk, targetLang);
        translatedChunks.push(translatedChunk);
      }

      // Восстанавливаем текст с сохранением ссылок
      const translatedText = restoreTextWithLinks(translatedChunks, links);

      return {
        text: translatedText,
        sourceLanguage: undefined, // Yandex определяет язык автоматически
        provider: this.provider,
      };
    } catch (error) {
      if (error instanceof TranslationError) {
        throw error;
      }
      throw this.createError(
        `Yandex Translate translation failed: ${error instanceof Error ? error.message : String(error)}`,
        undefined,
        error
      );
    }
  }

  /**
   * Переводит один чанк текста
   * @param chunk Текст для перевода
   * @param targetLang Целевой язык
   * @returns Переведённый текст
   */
  private async translateChunk(chunk: string, targetLang: string): Promise<string> {
    // Нормализация кодов языков для Yandex
    const normalizedLang = this.normalizeYandexLanguageCode(targetLang);

    // Генерируем уникальный идентификатор
    const ucid = uuidv4().replace(/-/g, '');

    // Формируем параметры запроса
    const params = new URLSearchParams();
    params.append('text', chunk);
    params.append('lang', normalizedLang);

    // Выполняем запрос - без API ключа, как в SillyTavern
    const result = await fetch(
      `${this.apiUrl}?ucid=${ucid}&srv=android&format=text`,
      {
        method: 'POST',
        body: params,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        signal: AbortSignal.timeout(this.timeout),
      }
    );

    if (!result.ok) {
      const errorText = await result.text();
      throw this.createError(
        `Yandex Translate API error: ${result.statusText}`,
        result.status,
        errorText
      );
    }

    const json = await result.json();

    if (json.code !== 200) {
      throw this.createError(
        `Yandex Translate API error: ${json.code} - ${json.message}`,
        json.code,
        json
      );
    }

    // Yandex возвращает массив переведённых текстов
    return json.text.join('');
  }

  /**
   * Нормализует код языка для Yandex Translate
   * @param lang Код языка
   * @returns Нормализованный код
   */
  private normalizeYandexLanguageCode(lang: string): string {
    // Специфичные преобразования для Yandex Translate
    const yandexLanguageMap: Record<string, string> = {
      'zh': 'zh',      // Yandex использует 'zh' для китайского
      'pt': 'pt',      // Yandex использует 'pt' для португальского
      'ru': 'ru',      // Русский
      'en': 'en',      // Английский
      'de': 'de',      // Немецкий
      'fr': 'fr',      // Французский
      'es': 'es',      // Испанский
      'it': 'it',      // Итальянский
      'ja': 'ja',      // Японский
      'ko': 'ko',      // Корейский
    };

    return yandexLanguageMap[lang] || lang;
  }

  /**
   * Определяет язык текста
   * @param text Текст для определения языка
   * @returns Код языка
   */
  async detectLanguage(text: string): Promise<string> {
    try {
      const params = new URLSearchParams();
      params.append('text', text);
      params.append('lang', 'auto');

      const result = await fetch(
        `${this.apiUrl}?ucid=${uuidv4().replace(/-/g, '')}&srv=android&format=text`,
        {
          method: 'POST',
          body: params,
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
          },
          signal: AbortSignal.timeout(this.timeout),
        }
      );

      if (!result.ok) {
        throw this.createError(
          `Yandex Translate API error: ${result.statusText}`,
          result.status
        );
      }

      const json = await result.json();
      return json.lang || 'unknown';
    } catch (error) {
      throw this.createError(
        `Language detection failed: ${error instanceof Error ? error.message : String(error)}`,
        undefined,
        error
      );
    }
  }

  /**
   * Получает список поддерживаемых языков
   * @returns Список языков
   */
  async getSupportedLanguages(): Promise<string[]> {
    // Yandex Translate поддерживает более 90 языков
    return [
      'af', 'sq', 'am', 'ar', 'hy', 'az', 'eu', 'be', 'bn', 'bs',
      'bg', 'ca', 'zh', 'hr', 'cs', 'da', 'nl', 'en', 'et', 'fi',
      'fr', 'ka', 'de', 'el', 'he', 'hi', 'hu', 'is', 'id', 'ga',
      'it', 'ja', 'kk', 'ko', 'lv', 'lt', 'mk', 'no', 'fa', 'pl',
      'pt', 'ro', 'ru', 'sr', 'sk', 'sl', 'es', 'sv', 'th', 'tr',
      'uk', 'ur', 'vi',
    ];
  }
}